### DISCUSSION TITLE
Help please! Fatal glibc error: malloc assertion failure in sysmalloc

### DISCUSSION THREAD

**Discussion by VLamek** [community/community#113322] [Programming Help]
### Body

My code throws an error that I have no idea how to fix. The error is in the insert function. I marked it with the comment with arrow. Here's the code: 
"Fatal glibc error: malloc assertion failure in sysmalloc: (old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)"

Here's the code:
`
#include <iostream>
using namespace std;

class Complex {
private:
  double real;
  double img;

public:
  Complex() {
    real = 0;
    img = 0;
  }

  Complex(double r, double i) : real(r), img(i) {}

  double getReal() { return real; }

  double getImg() { return img; }

  friend ostream &operator<<(ostream &os, Complex &complex) {

    if (complex.getImg() >= 0) {
      os << complex.getReal() << " + " << complex.getImg() << "i";
    }

    if (complex.getImg() < 0) {
      os << complex.getReal() << " - " << complex.getImg() * (-1) << "i";
    }

    return os;
  }

  class DinMas {
  private:
    Complex *Arr; // source array
    int Len;      

    int newSize;
    Complex *newArr;

    int newSize_1; 
    Complex *newArr_1; 

    int newSize_2;
    Complex *newArr_2; 

    int SIZE;
    Complex *ARR;

    int SIZE_insert;
    Complex *ARR_insert;

  public:
    int getSIZE() { return SIZE; }

    int getSIZE_insert() { return SIZE_insert; }

    Complex *getARR() { return ARR; }

    Complex &operator[](int index) {
      if (index < 0 || index >= Len) {
        throw std::out_of_range("The index is out of range");
      }
      return Arr[index];
    }

    DinMas() {
      Arr = NULL;
      Len = 0;
    }

    DinMas(int size) {
      Arr = new Complex[size];
      Len = size;
      std::cout << Len << endl;
      Complex zero(0, 0);

      for (int i = 0; i < size; i++) {
        Arr[i] = zero;
      }
    }

    double getLen() { return Len; }

    Complex *merge(Complex arr1[], int size1, Complex arr2[], int size2) {

      SIZE = size1 + size2; // creating the final array
      ARR = new Complex[SIZE];

      for (int i = 0; i < size1; i++) { // 1st to final
        ARR[i] = arr1[i];
      }

      for (int i = 0; i < SIZE; i++) { // 2nd to final
        ARR[size1 + i] = arr2[i];
      }

      std::cout << "Merged Array:" << std::endl;

      for (int i = 0; i < SIZE; i++) {
        std::cout << ARR[i] << std::endl;
      }

      std::cout << endl;

      return ARR;
    } // end of merge

    Complex *insert(Complex add) {
      std::cout << "Function started" << endl;
      SIZE_insert = Len + 1;
      std::cout << "Size : " << SIZE_insert << endl;
      newArr_1 = new Complex[SIZE_insert]; // <==== THE PROBLEM IS HERE
      std::cout << "Array created" << endl;

      for (int i = 0; i < Len; i++) {
        ARR_insert[i] = Arr[i];
      }

      ARR_insert[SIZE_insert - 1] = add;
      std::cout << "Inserted element test----> " << ARR_insert[SIZE_insert - 1] << endl << endl;
      return ARR_insert;
    }

    ~DinMas() {
      delete[] Arr;
      delete[] newArr;
      delete[] newArr_1;
      delete[] newArr_2;
      //delete[] ARR;
      delete[] ARR_insert;
    }
  };
};

int main() {

  // start of the merge test

  int wd = 8; // lengths for 2 test arrays
  int re = 15;

  std::cout << "Number of elements test----> " << re + wd << endl << endl;

  Complex *WD;
  Complex *RE;

  WD = new Complex[wd]; // creating test arrays
  RE = new Complex[re];

  Complex one(1, 0);
  Complex twoOnes(1, 1);

  for (int i = 0; i < wd; i++) { // filling in the arrays for clarity
    WD[i] = one;
  }

  for (int i = 0; i < re; i++) {
    RE[i] = twoOnes;
  }

  Complex::DinMas dinmas_merge;
  dinmas_merge.merge(WD, wd, RE, re);

  delete[] WD;
  delete[] RE;
  // end of the merge test

  // start of the insert test
  Complex element(123, 34); // this element will be inserted

  Complex::DinMas Array(12);
  Array.insert(element); // <======== the function does not work

  std::cout << "Array with an inserted element:" << endl;

  int leng = Array.getSIZE_insert();

  for (int i = 0; i < leng; i++) {
    std::cout << Array[i] << endl;
  }
  
  return 0;
}
`

How can i fix it?  :(

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by winstontsui** [comment]
In your insert function, when you allocate memory for newArr_1 you’re not doing it for the ARR_insert variable.

Try this:
Complex *insert(Complex add) {
std::cout << "Function started" << endl;
SIZE_insert = Len + 1;
std::cout << "Size : " << SIZE_insert << endl;
ARR_insert = new Complex[SIZE_insert]; // Allocate memory for ARR_insert
newArr_1 = new Complex[SIZE_insert];
std::cout << "Array created" << endl;

for (int i = 0; i < Len; i++) {
    ARR_insert[i] = Arr[i];
}

ARR_insert[SIZE_insert - 1] = add;
std::cout << "Inserted element test----> " << ARR_insert[SIZE_insert - 1] << endl << endl;
return ARR_insert;
}
